File: ch3/UVa00750.py
def canPlace(r, c):
    for prev in range(c):
        if row[prev] == r or abs(row[prev]-r) == abs(prev-c):
            return False
    return True

def backtrack(c):
    global lineCounter
    if c == 8 and row[b] == a:
        lineCounter += 1
        print("%2d     " % lineCounter, *[i+1 for i in row])
        return
    for r in range(8):
        if c == b and r != a:
            continue
        if canPlace(r, c):
            row[c] = r
            backtrack(c+1)

for tc in range(int(input())):
    if tc > 0:
        print()
    input()
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    row = [0] * 8
    lineCounter = 0
    print("SOLN       COLUMN")
    print(" #      1 2 3 4 5 6 7 8\n")
    backtrack(0)

File: ch3/cs/UVa00750.cpp
#include <bits/stdc++.h>
using namespace std;

int row[8], a, b, lineCounter;                   // global variables

bool canPlace(int r, int c) {
  for (int prev = 0; prev < c; ++prev)           // check previous Queens
    if ((row[prev] == r) || (abs(row[prev]-r) == abs(prev-c)))
      return false;                              // infeasible
  return true;
}

void backtrack(int c) {
  if ((c == 8) && (row[b] == a)) {               // a candidate sol
    printf("%2d      %d", ++lineCounter, row[0]+1);
    for (int j = 1; j < 8; ++j) printf(" %d", row[j]+1);
    printf("\n");
    return;                                      // optional statement
  }
  for (int r = 0; r < 8; ++r) {                  // try all possible row
    if ((c == b) && (r != a)) continue;          // early pruning
    if (canPlace(r, c))                          // can place a Queen here?
      row[c] = r, backtrack(c+1);                // put here and recurse
  }
}

int main() {
  int TC; scanf("%d", &TC);
  while (TC--) {
    scanf("%d %d", &a, &b); --a; --b;            // to 0-based indexing
    memset(row, 0, sizeof row); lineCounter = 0;
    printf("SOLN       COLUMN\n");
    printf(" #      1 2 3 4 5 6 7 8\n\n");
    backtrack(0);                                // sub 8! operations
    if (TC) printf("\n");
  }
  return 0;
}
